lightning_examples/finetuning-scheduler/finetuning-scheduler.py

```diff
@@ -0,0 +1,565 @@
+                    for n, p in self.model.named_parameters()
+                    if any(nd in n for nd in self.no_decay) and p.requires_grad
+                ],
+                "weight_decay": 0.0,
+            },
+        ]
+        return pgs
+
+    def configure_optimizers(self):
+        # the phase 0 parameters will have been set to require gradients during setup
+        # you can initialize the optimizer with a simple requires.grad filter as is often done,
+        # but in this case we pass a list of parameter groups to ensure weight_decay is
+        # not applied to the bias parameter (for completeness, in this case it won't make much
+        # performance difference)
+        optimizer = instantiate_registered_class(args=self.init_pgs(), init=self.optimizer_init)
+        scheduler = {
+            "scheduler": instantiate_registered_class(args=optimizer, init=self.lr_scheduler_init),
+            **self.pl_lrs_cfg,
+        }
+        return [optimizer], [scheduler]
+
+    def configure_callbacks(self):
+        found_fts = [c for c in self.trainer.callbacks if isinstance(c, FinetuningScheduler)]  # type: ignore # noqa
+        if found_fts:
+            self.finetuningscheduler_callback = found_fts[0]
+        return super().configure_callbacks()
+
+
+# %%
+# let's create a finetuning schedule for our model and run an explicitly scheduled finetuning training scenario with it
+# Please see the documentation for a full description of the schedule format
+ft_schedule_yaml = """
+0:
+  params: # the parameters for each phase definition can be fully specified
+  - model.classifier.bias
+  - model.classifier.weight
+1:
+  params: # or specified via a regex
+  - model.albert.pooler.*
+2:
+  params:
+  - model.albert.encoder.*.ffn_output.*
+"""
+ft_schedule_name = "RteBoolqModule_ft_schedule_albert_base.yaml"
+with open(ft_schedule_name, "w") as f:
+    f.write(ft_schedule_yaml)
+
+# %%
+datasets.set_progress_bar_enabled(False)
+pl.seed_everything(42)
+dm = RteBoolqDataModule(model_name_or_path="albert-base-v2", tokenizers_parallelism=False)
+dm.setup("fit")
+optimizer_init = {
+    "class_path": "torch.optim.AdamW",
+    "init_args": {"weight_decay": 1e-05, "eps": 1e-07, "lr": 1e-05},
+}
+lr_scheduler_init = {
+    "class_path": "torch.optim.lr_scheduler.CosineAnnealingWarmRestarts",
+    "init_args": {"T_0": 1, "T_mult": 2, "eta_min": 1e-07},
+}
+pl_lrs_cfg = {"interval": "epoch", "frequency": 1, "name": "CosineAnnealingWarmRestarts"}
+
+model = RteBoolqModule(
+    model_name_or_path="albert-base-v2",
+    optimizer_init=optimizer_init,
+    lr_scheduler_init=lr_scheduler_init,
+    pl_lrs_cfg=pl_lrs_cfg,
+    experiment_tag="fts_explicit",
+)
+callbacks = [
+    FinetuningScheduler(ft_schedule=ft_schedule_name, max_depth=2),  # type: ignore # noqa
+    FTSEarlyStopping(monitor="val_loss", min_delta=0.001, patience=2),  # type: ignore # noqa
+    FTSCheckpoint(monitor="val_loss", save_top_k=5),  # type: ignore # noqa
```

Comment:
Ideally, this shouldn't be necessary.

Reply:
The FinetuningScheduler will bring in the two other required callback extensions ``FTSEarlyStopping`` and ``FTSCheckpoint`` with default configurations if I don't include them. I've just included them here to be explicit and demonstrate how to set non-default configurations for them but if you think it's clearer not to include them I can let the default ones be added.

Reply:
Mind adding a comment above to explain this?